core/src/main/java/io/grpc/transport/ServerTransportListener.java

```diff
@@ -48,4 +49,9 @@
    */
   ServerStreamListener streamCreated(ServerStream stream, String method,
       Metadata.Headers headers);
+
+  /**
+   * The transport completed shutting down. All resources have been released.
+   */
+  void transportTerminated();
```

Comment:
> It doesn't answer the question users will be asking. This means instead of doing if (isShutdown()) like you do now, you have to do State state = state(); if (state == SHUTTING_DOWN || state == TERMINATED). 

I don't see this as a big problem. If you wanted to make the caller's life easier you could give the enum a utility method to help with common checks such as `state.isShutdownOrTerminated`.

> You can't add new states in the future, because of point 1. That means you either err on the side of making too many states (making point 1 worse) or you err on the side of too few and now you have the same problem we have now. This is quite worrisome.

I don't see us needing more states ever.  If our lifecycle management changes drastically enough for us to warrant new states, it will likely be an API breaker anyway. I don't think the likelihood of this happening is enough to warrant the concern.

I think a single state makes the API way more simplified and understandable ... and IMHO that outweighs the cons in this case.